eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs

```diff
@@ -397,4 +272,282 @@ impl Eth2NearRelay {
+        trace!(target: "relay", "Init slot for search as {}", slot);
+
+        let last_eth_slot = self.beacon_rpc_client.get_last_slot_number()?.as_u64();
+
+        return if self.enable_binsearch {
+            self.binary_slot_search(slot, finalized_slot, last_eth_slot)
+        } else {
+            self.linear_slot_search(slot, finalized_slot, last_eth_slot)
+        };
+    }
+
+    fn linear_slot_search(
+        &self,
+        slot: u64,
+        finalized_slot: u64,
+        last_eth_slot: u64,
+    ) -> Result<u64, Box<dyn Error>> {
+        return if slot == finalized_slot || self.block_known_on_near(slot)? {
+            Ok(self.linear_search_forward(slot, last_eth_slot))
+        } else {
+            Ok(self.linear_search_backward(finalized_slot, slot))
+        };
+    }
+
+    fn binary_slot_search(
+        &self,
+        slot: u64,
+        finalized_slot: u64,
+        last_eth_slot: u64,
+    ) -> Result<u64, Box<dyn Error>> {
+        return if slot == finalized_slot || self.block_known_on_near(slot)? {
+            self.binsearch_slot_forward(slot, last_eth_slot + 1)
+        } else {
+            self.binsearch_slot_range(finalized_slot, slot)
+        };
+    }
+
+    fn binsearch_slot_forward(&self, slot: u64, max_slot: u64) -> Result<u64, Box<dyn Error>> {
+        let mut current_step = 1;
+        let mut prev_slot = slot;
+        while slot + current_step < max_slot {
+            match self.block_known_on_near(slot + current_step) {
+                Ok(true) => {
+                    prev_slot = slot + current_step;
+                    current_step = min(current_step * 2, max_slot - slot);
+                }
+                Ok(false) => break,
+                Err(_) => {
+                    let (slot_id, slot_on_near) =
+                        self.find_left_non_error_slot(slot + current_step, max_slot);
+                    if slot_on_near {
+                        prev_slot = slot_id;
+                        current_step = min(current_step * 2, max_slot - slot);
+                    } else {
+                        current_step = slot_id - slot;
+                        break;
+                    }
+                }
+            }
+        }
+
+        self.binsearch_slot_range(prev_slot, slot + current_step)
+    }
+
+    fn binsearch_slot_range(&self, start_slot: u64, last_slot: u64) -> Result<u64, Box<dyn Error>> {
+        let mut start_slot = start_slot;
+        let mut last_slot = last_slot;
+        while start_slot + 1 < last_slot {
+            let mid_slot = start_slot + (last_slot - start_slot) / 2;
+            match self.block_known_on_near(mid_slot) {
+                Ok(true) => start_slot = mid_slot,
+                Ok(false) => last_slot = mid_slot,
+                Err(_) => {
+                    let (left_slot, is_left_slot_on_near) =
+                        self.find_left_non_error_slot(mid_slot, last_slot);
+                    if is_left_slot_on_near == true {
```

Comment:
Redundant comparison

```suggestion
                    if is_left_slot_on_near {
```